src/compiler/transformers/type-library.ts

```diff
@@ -0,0 +1,298 @@
+import { normalizePath } from '@utils';
+import path from 'path';
+import ts from 'typescript';
+
+import type * as d from '../../declarations';
+import { typeToString } from './transform-utils';
+
+/**
+ * This is a {@link d.JsonDocsTypeLibrary} cache which is used to store a
+ * record of all the types referenced by components which are 'seen' during a
+ * Stencil build
+ */
+const TYPE_LIBRARY: d.JsonDocsTypeLibrary = {};
+
+/**
+ * Add a type reference to the library if it has not already been added. This
+ * function then returns the unique identifier for that type so a reference to
+ * it can be stored.
+ *
+ * @param type the type we want to add
+ * @param typeName the type's name
+ * @param checker a {@link ts.TypeChecker} instance
+ * @param pathToTypeModule the path to the home module of the type
+ * @returns the unique ID for the type in question
+ */
+export function addToLibrary(
+  type: ts.Type,
+  typeName: string,
+  checker: ts.TypeChecker,
+  pathToTypeModule: string
+): string {
+  pathToTypeModule = path.relative(process.cwd(), normalizePath(pathToTypeModule)).replace(/^\.\//, '');
+
+  // for now we don't make any attempt to include types in node_modules
+  if (pathToTypeModule.startsWith('node_modules')) {
+    return '';
+  }
+
+  const id = getTypeId(pathToTypeModule, typeName);
+
+  if (!(id in TYPE_LIBRARY) && !type.isTypeParameter()) {
+    const declaration = getTypeDeclaration(checker, type);
+
+    if (declaration !== '') {
+      TYPE_LIBRARY[id] = {
+        declaration,
+        docstring: getTypeDocstring(type, checker),
+        path: pathToTypeModule,
+      };
+    }
+  }
+
+  return id;
+}
+
+/**
+ * This returns a reference to the cached {@link d.JsonDocsTypeLibrary} which
+ * lives in this module.
+ *
+ * @returns a reference to the type library
+ */
+export function getTypeLibrary(): d.JsonDocsTypeLibrary {
+  return TYPE_LIBRARY;
+}
+
+/**
+ * Helper function that, given a file containing interfaces to document, will
+ * add all the types exported from that file to the type library.
+ *
+ * This will exclude any types that are marked 'private' via JSDoc.
+ *
+ * @param filePath the path to the file of interest
```

Comment:
For the purposes of this file the file path only needs to be resolvable from `process.cwd()`, so that could be a relative or absolute path. I could add a note to that effect.